core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -144,58 +162,22 @@
     private final TableExecuteContextManager tableExecuteContextManager;
     private final SplitSourceFactory splitSourceFactory;
 
-    private final Map<PlanFragmentId, SqlStageExecution> stages;
-    private final StageId rootStageId;
-    private final Map<StageId, Set<StageId>> stageLineage;
+    private final StageExecutions stages;
+    private final CoordinatorStagesScheduler coordinatorStagesScheduler;
 
-    // all stages that could be scheduled remotely (excluding coordinator only stages)
-    private final Set<StageId> remotelyScheduledStages;
+    private final RetryPolicy retryPolicy;
+    private final int maxRetryAttempts;
+    private final AtomicInteger currentAttempt = new AtomicInteger();
+    private final Duration retryDelay;
 
     private final AtomicBoolean started = new AtomicBoolean();
-    private final AtomicReference<PipelinedScheduler> scheduler = new AtomicReference<>();
 
-    public static SqlQueryScheduler createSqlQueryScheduler(
-            QueryStateMachine queryStateMachine,
-            SubPlan plan,
-            NodePartitioningManager nodePartitioningManager,
-            NodeScheduler nodeScheduler,
-            RemoteTaskFactory remoteTaskFactory,
-            boolean summarizeTaskInfo,
-            int splitBatchSize,
-            ExecutorService queryExecutor,
-            ScheduledExecutorService schedulerExecutor,
-            FailureDetector failureDetector,
-            NodeTaskMap nodeTaskMap,
-            ExecutionPolicy executionPolicy,
-            SplitSchedulerStats schedulerStats,
-            DynamicFilterService dynamicFilterService,
-            TableExecuteContextManager tableExecuteContextManager,
-            Metadata metadata,
-            SplitSourceFactory splitSourceFactory)
-    {
-        SqlQueryScheduler sqlQueryScheduler = new SqlQueryScheduler(
-                queryStateMachine,
-                plan,
-                nodePartitioningManager,
-                nodeScheduler,
-                remoteTaskFactory,
-                summarizeTaskInfo,
-                splitBatchSize,
-                queryExecutor,
-                schedulerExecutor,
-                failureDetector,
-                nodeTaskMap,
-                executionPolicy,
-                schedulerStats,
-                dynamicFilterService,
-                tableExecuteContextManager,
-                metadata,
-                splitSourceFactory);
-        sqlQueryScheduler.initialize();
-        return sqlQueryScheduler;
-    }
+    @GuardedBy("this")
+    private final AtomicReference<DistributedStagesScheduler> distributedStagesScheduler = new AtomicReference<>();
```

Comment:
You may consider changing this to `AtomicReference<Optional<DistributedStagesScheduler>>` and then 
in `queryStateMachine.addStateChangeListener(state -> {` you may checkState that value read from `this.distributedStagesScheduler.get();` is not-null (that atomic reference was actually set, just maybe to Optional.empty). A sanity check.

Reply:
I agree that it might be a good validation to have specifically in the `queryStateMachine.addStateChangeListener ...` code. However in other places `DistributedStagesScheduler` may legitimately be not yet created when something happens and having extra Optional checks seem to be an unnecessary overhead. Given that the scope around `queryStateMachine.addStateChangeListener ...` is rather narrow it feels like it is rather better to keep it as it is.